### PR TITLE
Expose a TypeMarker factory to wrap an arbitrary Type

### PR DESCRIPTION
**Pull request by carterkozak**
This mirrors a change we've just introduced to conjure-undertow here: https://github.com/palantir/conjure-java/pull/2264

Dialogue takes advantage of this new factory method in ConjureBodySerDe to avoid holding class references in serializer caches, which introduced some friction in our dns-polling lifecycle management. I haven't reverted the change to specify `weakKeys` in the caches, however the keys should never be freed unless classes are unloaded (e.g. plugins).

==COMMIT_MSG==
Expose a TypeMarker factory to wrap an arbitrary Type
==COMMIT_MSG==

### CODE REVIEW

File: dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java [hunk 1]
```diff
@@ -112,13 +113,13 @@ private ImmutableList<Encoding> sortByWeight(List<WeightedEncoding> encodings) {
...
     public <T> Serializer<T> serializer(TypeMarker<T> token) {
-        return (Serializer<T>) serializers.get(token);
+        return (Serializer<T>) serializers.get(token.getType());
     }
 
     @Override
     @SuppressWarnings("unchecked")
     public <T> Deserializer<T> deserializer(TypeMarker<T> token) {
-        return (Deserializer<T>) deserializers.get(token);
+        return (Deserializer<T>) deserializers.get(token.getType());
```
Comment (carterkozak) [inline dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/ConjureBodySerDe.java:122]: Changes to this file would also make sense in a separate PR

**Review by pkoenig10 — APPROVED** [review]
